tests/src/test/scala/cats/collections/HashMapSuite.scala

```diff
@@ -164,6 +165,18 @@ class HashMapSuite extends DisciplineSuite with ScalaVersionSpecificSyntax {
     }
   }
 
+  test("removed with collision") {
+    val myHash: Hash[Int] = new Hash[Int] {
```

Comment:
one idea I was to make a generator for `Hash[Int]` which has two functions `g: Int => Int` and `h: Int => Int`

and:
```scala
new Hash[Int] {
  def eqv(x: Int, y: Int) = g(x) == g(y)
  def hash(x: Int) = h(g(x))
}
```

and use that `Hash` with our generative tests. Hopefully that would have actually found the previous bug. I'm a bit worried we are stabbing in the dark here.

Reply:
How about something like this:

```scala
  def genCollidingKey(hm: HashMap[String, Int]): Gen[String] =
    if (hm.isEmpty)
      Arbitrary.arbitrary[String]
    else
      Gen.oneOf(hm.keysIterator.toList).map(collidingString)

  test("removed with collision") {
    forAll { (initial: HashMap[String, Int], value: Int) =>
      forAll(genCollidingKey(initial)) { (key: String) =>
        val colliding = initial.updated(key, value)
        val removed = colliding.removed(key)
        assert(!removed.contains(key))
        assert(removed.get(key) === None)
        if (initial.contains(key))
          assert(removed.size === (initial.size - 1))
        else
          assert(removed === initial)
      }
    }
  }
```
This fails with @durban's code change reverted and succeeds after it.

However, it still doesn't test removal with multiple colliding keys.

I think the ideal would be to inject collisions into the generation process, since so much of the testing here relies on the `hashMapGen` using all of the primitive operations in order to generate the initial values for testing.

Perhaps we could add a specialised the `hashMapGen` for `String` or `Int` in order to generate frequent collisions as part of generating the initial test values?